### ISSUE TITLE
Shard awareness - rare misrouting of connection

### ISSUE THREAD

**Issue by jul-stas** [scylladb/scylladb#7277]
On a fresh master, e.g. a2cc599a2a026e4cc07eb61653ff8a42af4e490e, experimental advanced-shard-aware cpp-driver said:
```
[WARN] (connector.cpp:305:void datastax::internal::core::Connector::on_supported(datastax::internal::core::ResponseMessage*)): Connected to 127.0.0.1:19042 to shard 3 from local port 50008, but expected shard 0. Is client behind a NAT?
```
There are 4 shards, so 50008 should connect to shard 0. There is no NAT.
It hapens rarely, once in a few hundred connections.

EDIT: AFAIK drivers are resilient to this, because dealing with NAT is a standard procedure. E.g. cpp-driver is unaffected, but the origin of this error must be investigated.

**Comment by slivne** [comment]
isn't routing based on the destination port in scylla ?  so is int 19042 that defined the shard to be connected to not 50008 ?

**Comment by slivne** [comment]
We have to fix this bug or shardawarness will not work for customers in the field with the new drivers

**Comment by jul-stas** [comment]
> isn't routing based on the destination port in scylla ? so is int 19042 that defined the shard to be connected to not 50008 ?

Destination port 19042 pays attention to the client-side port. So it should connect the client to shard `50008 % 4 == 0`.
We decided to do it this way to avoid opening too many server-side ports.

**Comment by slivne** [comment]
@jul-stas ping

**Comment by haaawk** [comment]
What's the question @slivne ?

**Comment by slivne** [comment]
where is the fix :)

**Comment by jul-stas** [comment]
I'll have a look, it's not trivial to reproduce.

**Comment by jul-stas** [comment]
This turns out to be a race in cpp-driver: "scout connectors" (non shard-aware) can find spurious `sharding_info` and try to connect on port 9042 expecting source-port-based-routing there. `system.clients` table does not distinguish destination ports, hence the confusion.
Scylla is OK. Driver's function is also fine, because "scout connections" die soon after, but these warnings need to be fixed.
Closing!

**Comment by vladzcloudius** [comment]
@jul-stas was there a fix? I don't see any reference. If there wasn't a fix - please, don't close there issue till there is.

The same issue has just been reported for a Java driver `3.11.2.1`

**Comment by jul-stas** [comment]
> @jul-stas was there a fix?

No. But it's NAB - just a transient warning; and cpp-driver is being replaced by cpp-rust-wrapper so I'm hesitant to introduce any risky changes. Let's get the opinion from the master of drivers, CC @avelanarius

**Comment by avelanarius** [comment]
> The same issue has just been reported for a Java driver `3.11.2.1`

@vladzcloudius Could you point me to where it was reported for Java Driver?

**Comment by vladzcloudius** [comment]
> > The same issue has just been reported for a Java driver `3.11.2.1`
> 
> @vladzcloudius Could you point me to where it was reported for Java Driver?

One of our customers reported it, @avelanarius.

**Comment by roydahan** [comment]
@avelanarius is this issue still relevant?

**Comment by avelanarius** [comment]
Closing this old issue - there are no details on why this happens, there's no reproducer and no one reported that this issue happened to them recently.

**Comment by maverick-ai** [comment]
It is happening to us in gocql

**Comment by roydahan** [comment]
> It is happening to us in gocql when inside docker

Can you please provide more details?
- gocql version
- scylladb version

**Comment by maverick-ai** [comment]
Scylla Version : 5.4 open source
scylla variant of gocql: latest

The issue is not prevailing when it is outside docker. We even tried network_mode: "host"

**Comment by maverick-ai** [comment]
@roydahan Any update on this?

**Comment by roydahan** [comment]
> @roydahan Any update on this?

Not yet, but it's on the team's queue.

**Comment by roydahan** [comment]
Let's start by checking why this issue isn't reproduced in our integration tests.

**Comment by sylwiaszunejko** [comment]
> It is also very easy to reproduce.

@maverick-ai Could you provide exact steps to reproduce the issue?

**Comment by sylwiaszunejko** [comment]
@maverick-ai ping

**Comment by sylwiaszunejko** [comment]
@maverick-ai ping

**Comment by maverick-ai** [comment]
The issue is reproducible on docker instance of ScyllaDB. It auto disconnects in few days. When we installed on Server directly, the problem disappeared

**Comment by maverick-ai** [comment]
It occurs randomly. But mostly it occurs when running inside docker.

**Comment by dkropachev** [comment]
We have tried to reproduce it with steps given, this error is not happening.
@maverick-ai , can you please give more details on how exactly do you reproduce it.

**Comment by maverick-ai** [comment]
We created a cluster of 2 machines on digital ocean droplet running on Ubuntu 22.04 one Scylla instance on each machine. When we try to connect to the Scylla instance from our office. We received this message. The message occurs when we try to connect to Scylla instance from our docker golang container. This error message occurs randomly, sometime it occurs sometimes not. It also occurs when we try to connect from other droplets.

**Comment by dkropachev** [comment]
How docker golang container is getting routed to the digital ocean instances ?

**Comment by maverick-ai** [comment]
Docker container is running inside Digital Ocean droplet

**Comment by dkropachev** [comment]
> Digital Ocean droplet

It does not answer the question, can you please provide network description.
Considering the fact that you mentioned that you connecting from the office I would suggest that you have nat there and connection to the scylla cluster goes thrue that nat.

**Comment by maverick-ai** [comment]
@dkropachev there is no nat in my office. I tried even from the mobile hotspot.  We are using digital ocean droplet that are basically ubuntu VMs. We install docker and run our API server. If run the same code by go run main.go inside the droplet the error does not occur. I don't know how because of docker, the driver is creating the issue.

**Comment by dkropachev** [comment]
> @dkropachev there is no nat in my office. I tried even from the mobile hotspot. We are using digital ocean droplet that are basically ubuntu VMs. We install docker and run our API server. If run the same code by go run main.go inside the droplet the error does not occur. I don't know how because of docker, the driver is creating the issue.

Can you please answer this question https://github.com/scylladb/scylladb/issues/7277#issuecomment-2273508622

**Comment by dkropachev** [comment]
This is expected behavior in docker setup when `bridge` is used, because docker uses `SNAT`.
So driver can't fulfill all the shard connections right away and then uses any available connection to the host to run the query.
